src/udp_mux/mod.rs

```diff
@@ -0,0 +1,332 @@
+                    }
+                    Err(err) if err.kind() == ErrorKind::TimedOut => continue,
+                    Err(err) => {
+                        log::error!("Could not read udp packet: {}", err);
+                        break;
+                    }
+                };
+            }
+        });
+    }
+}
+
+#[async_trait]
+impl UDPMux for UDPMuxDefault {
+    async fn close(&self) {
+        if self.is_closed() {
+            return;
+        }
+
+        self.closed.store(true, Ordering::SeqCst);
+
+        let old_conns = {
+            let mut conns = self.conns.lock().await;
+
+            std::mem::take(&mut (*conns))
+        };
+
+        // NOTE: We don't wait for these closure to complete
+        for (_, conn) in old_conns.into_iter() {
+            conn.close();
+        }
+
+        {
+            let mut address_map = self
+                .address_map
+                .write()
+                .expect("Failed to obtain address_map lock");
+
+            // NOTE: This is important, we need to drop all instances of `UDPMuxConn` to
+            // avoid a retain cycle due to the use of [`std::sync::Arc`] on both sides.
+            let _ = std::mem::take(&mut (*address_map));
+        }
+    }
+
+    async fn get_conn(self: Arc<Self>, ufrag: &str) -> Result<Arc<dyn Conn + Send + Sync>, Error> {
+        if self.is_closed() {
+            return Err(Error::ErrUseClosedNetworkConn);
+        }
+
+        {
+            let mut conns = self.conns.lock().await;
+            if let Some(conn) = conns.get(ufrag) {
+                // UDPMuxConn uses `Arc` internally so it's cheap to clone, but because
+                // we implement `Conn` we need to further wrap it in an `Arc` here.
+                return Ok(Arc::new(conn.clone()) as Arc<dyn Conn + Send + Sync>);
+            }
+
+            let muxed_conn = self.create_muxed_conn(ufrag).await;
+            let mut close_rx = muxed_conn.close_rx();
+            let cloned_self = Arc::clone(&self);
+            let cloned_ufrag = ufrag.to_string();
+            tokio::spawn(async move {
+                let _ = close_rx.changed().await;
+
+                // Arc needed
+                cloned_self.remove_conn_by_frag(&cloned_ufrag).await;
+            });
+
+            conns.insert(ufrag.into(), muxed_conn.clone());
+
+            Ok(Arc::new(muxed_conn) as Arc<dyn Conn + Send + Sync>)
+        }
+    }
+
+    async fn remove_conn_by_frag(&self, ufrag: &str) {
+        // Pion's ice implementation has both `RemoveConnByFrag` and `RemoveConn`, but since `conns`
+        // is keyed on `ufrag` their implementation is equivalent.
+
+        let removed_conn = {
+            let mut conns = self.conns.lock().await;
+            conns.remove(ufrag)
+        };
+
+        if let Some(conn) = removed_conn {
+            let mut address_map = self
+                .address_map
+                .write()
+                .expect("Failed to obtain write lock for address_map");
```

Comment:
no expect() to avoid panic. Add Result<()> as remove_conn_by_frag fn's return value.

Reply:
I switche do the sync primitives from `util`